File: 05-binary_search/00875-Koko_Eating_Bananas/00875-koko_eating_bananas.cpp
// 875: Koko Eating Bananas
// https://leetcode.com/problems/koko-eating-bananas/


#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


class Solution {
public:
    // SOLUTION
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = *max_element(piles.begin(), piles.end());
        
        while (l < r) {
            int total = 0;
            int m = (l + r) / 2;

            for (int p : piles)
                total += (p + m - 1) / m;
            if (total > h)
                l = m + 1;
            else
                r = m;
        }

        return l;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> piles = {3,6,7,11};
    int h = 8;

    // OUTPUT
    auto result = o.minEatingSpeed(piles, h);
    cout<<result<<endl;

    return 0;
}

File: 05-binary_search/00875-Koko_Eating_Bananas/00875-koko_eating_bananas.go
// 875: Koko Eating Bananas
// https://leetcode.com/problems/koko-eating-bananas/


package main

import "fmt"

// SOLUTION
func minEatingSpeed(piles []int, h int) int {
    l, r := 1, int(10e9)

	for l < r {
		total := 0
		m := (l + r) / 2

		for _, p := range piles {
			total += (p + m - 1) / m
		}
		if (total > h) {
			l = m + 1
		} else {
			r = m
		}
	}

	return l
}

func main() {
	// INPUT
	piles := []int{3,6,7,11}
	h := 8

	// OUTPUT
	result := minEatingSpeed(piles, h)
	fmt.Println(result)
}